### PR TITLE
Add -max_connections flag to put a hard limit on the number of connections

### PR DESCRIPTION
**Pull request by pdecat**
This is a first shot at trying to implement a hard limit on the number of connections to accept before refusing new connections to avoid OOM when memory resources are limited, as suggested by @Carrotman42 in #103.

TODO:

- [x] Add connections counter with atomic increments
- [x] Unit test
- [x] End to end test

### CODE REVIEW

File: proxy/proxy/client.go [hunk 1]
```diff
@@ -91,6 +95,11 @@ type cacheEntry struct {
 // proxy them to the destination instance. It blocks until connSrc is closed.
 func (c *Client) Run(connSrc <-chan Conn) {
 	for conn := range connSrc {
+		if c.MaxConnections > 0 && len(c.Conns.m) >= c.MaxConnections {
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: You need to check the number of connections in the entire map, not just the `len(c.Conns.m)` (which is the number of unique "id"s in the map); also you need to hold the lock in the ConnSet or else there's going to be a data race here. (we shouldn't read the unexported 'm' variable in this code anyway, even though it's allowed by the compiler because it happens to be in the same package.)

Even if you held the lock here or wrote a method on ConnSet to get the number of connections in the set, there's a logical race between checking the number of connections here and adding the new connection to the ConnSet in handleConn. Therefore I think this check needs to happen in handleConn.

Unfortunately I think this is too inefficient to work well anyway since you're adding an iteration through the ConnSet map on each new connection request, not to mention the inefficiency of ConnSet in the first place: I'm already hoping to remove a lot of the ConnSet logic from this Client because it is unfortunately pretty inefficient already (some `O(n)` operations which I completely forgot about and extra locking which probably reduces parallelism a little more than we'd like).

Barring the extra work to optimize the ConnSet, it's probably easiest to use a counter variable using atomic increments in this Client type to make this happen. That won't add much latency to new connections and avoids undesired coupling to the ConnSet. We can always move where that counter is in the future (it sort of belongs in the ConnSet, but would require some new methods on that type to make it work well).

Comment (pdecat) [inline proxy/proxy/client.go]: Great feedback, thanks!
I knew what I implemented was very far from perfect, but was kind of lost where to put the connection limit check.

At first, I wanted to place it in the Accept loop but I did not have access to the number of actual connections there.

I'll look into adding a counter like you suggest.

File: proxy/proxy/client.go [hunk 2]
```diff
@@ -36,6 +37,10 @@ const (
 // user.
 var errNotCached = errors.New("instance was not found in cache")
 
+// connectionsCounter is used to enforce the optional maxConnections limit
+var connectionsCounter uint64 = 0
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: This should be a field inside the Client struct instead of a package-level constant.

File: proxy/proxy/client.go [hunk 3]
```diff
@@ -109,6 +109,27 @@ func (c *Client) Run(connSrc <-chan Conn) {
...
+		connectionsCounterL.Lock()
+		if atomic.LoadUint64(&connectionsCounter) >= c.MaxConnections {
+			logging.Errorf("Too many open connections (max %d)", c.MaxConnections)
+			conn.Conn.Close()
+			connectionsCounterL.Unlock()
+			return
+		}
+
+		// Increment connectionsCounter then release exclusive lock
+		atomic.AddUint64(&connectionsCounter, 1)
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: Don't use a lock plus atomics: you can use the return value from the AddUint64 to tell if this increment tipped over the counter.

    active := atomic.AddUint64(&connectionsCounter, 1)
    defer atomic.AddUint64(&connectionsCounter, ^uint64(0))
    if active > c.MaxConnections {
       conn.Conn.Close()
       logging.Errorf("Too many open connections (max %d)", c.MaxConnections)
    }

File: proxy/proxy/client.go (hunk 2 above)
Comment (pdecat) [inline proxy/proxy/client.go]: Done, thanks!

File: proxy/proxy/client.go (hunk 3 above)
Comment (pdecat) [inline proxy/proxy/client.go]: Nice, thanks!

**Comment by pdecat** [comment]
Requested changes applied.

File: proxy/proxy/client.go [hunk 4]
```diff
@@ -114,23 +113,15 @@ func (c *Client) Run(connSrc <-chan Conn) {
...
+		// Deferred decrement of ConnectionsCounter upon connection closing
+		defer atomic.AddUint64(&c.ConnectionsCounter, ^uint64(0))
 
-		// Deferred decrement of connectionsCounter upon connection closing
-		defer func() {
-			atomic.AddUint64(&connectionsCounter, ^uint64(0))
-		}()
+		if active > c.MaxConnections {
+			conn.Conn.Close()
+			logging.Errorf("Too many open connections (max %d)", c.MaxConnections)
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: We need a return here

Comment (pdecat) [inline proxy/proxy/client.go]: Oops.

**Comment by Carrotman42** [comment]
The code looks good as-is. Can you add some testing as I referenced in the previous post?

File: cmd/cloud_sql_proxy/proxy.go [hunk 5]
```diff
@@ -292,7 +292,7 @@ func parseInstanceConfigs(dir string, instances []string, cl *http.Client) ([]in
 // CreateInstanceConfigs verifies that the parameters passed to it are valid
 // for the proxy for the platform and system and then returns a slice of valid
 // instanceConfig.
-func CreateInstanceConfigs(dir string, useFuse bool, instances []string, instancesSrc string, cl *http.Client) ([]instanceConfig, error) {
+func CreateInstanceConfigs(dir string, useFuse bool, instances []string, instancesSrc string, maxConnections uint64, cl *http.Client) ([]instanceConfig, error) {
```
Comment (Carrotman42) [inline cmd/cloud_sql_proxy/proxy.go]: Actually, just noticed: it doesn't appear that you need to modify this function anymore. Can you revert this change to this file? (unless I'm misreading)

Comment (pdecat) [inline cmd/cloud_sql_proxy/proxy.go]: You're right, fixed!

**Comment by pdecat** [comment]
I'm implementing some unit tests right now.

**Comment by pdecat** [comment]
Added a unit test but it fails from time to time.

```
2017/08/04 12:25:06 too many open connections (max 10)
2017/08/04 12:25:06 couldn't connect to "instance-name-0": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-3": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-8": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-10": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-4": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-1": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-5": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-2": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-6": this error is returned by the dialer
2017/08/04 12:25:06 couldn't connect to "instance-name-7": this error is returned by the dialer
2017/08/04 12:25:06 client has correctly refused to dial new connection on 11th attempt when the maximum of 10 connections was reached (10 dials)
PASS
ok      github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy       0.008s
```

```
2017/08/04 12:25:07 couldn't connect to "instance-name-10": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-0": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-1": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-2": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-3": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-4": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-5": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-6": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-7": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-8": this error is returned by the dialer
2017/08/04 12:25:07 couldn't connect to "instance-name-9": this error is returned by the dialer
--- FAIL: TestMaximumConnectionsCount (0.00s)
        client_test.go:174: client should have refused to dial new connection on 11th attempt when the maximum of 10 connections was reached (11 dials)
FAIL
exit status 1
FAIL    github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy       0.008s
```

**Comment by pdecat** [comment]
The instability of the unit test was caused by decrements occurring too soon after the dialer returned an error. Replacing the dialer lock by a 1 second pause fixed the issue.

File: proxy/proxy/client.go [hunk 6]
```diff
@@ -80,6 +81,14 @@ type Client struct {
...
 	cfgCache map[string]cacheEntry
 	cfgL     sync.RWMutex
+
+	// MaxConnections is the maximum number of connections to establish
+	// before refusing new connections. 0 means no limit.
+	MaxConnections uint64
+
+	// ConnectionsCounter is used to enforce the optional maxConnections limit
+	ConnectionsCounter  uint64
+	ConnectionsCounterL sync.RWMutex
```
Comment (Carrotman42) [inline proxy/proxy/client.go]: This is unused

File: proxy/proxy/client_test.go [hunk 7]
```diff
@@ -122,3 +124,59 @@ func TestConcurrentRefresh(t *testing.T) {
...
+		}(instanceName)
+	}
+
+	wg.Wait()
+
+	switch {
+	case dials > maxConnections:
+		t.Errorf("client should have refused to dial new connection on %dth attempt when the maximum of %d connections was reached (%d dials)", numConnections, maxConnections, dials)
+	case dials == maxConnections:
+		logging.Infof("client has correctly refused to dial new connection on %dth attempt when the maximum of %d connections was reached (%d dials)\n", numConnections, maxConnections, dials)
```
Comment (Carrotman42) [inline proxy/proxy/client_test.go]: use t.Logf since the log message is related to the test.

File: proxy/proxy/client_test.go [hunk 8]
```diff
@@ -122,3 +124,59 @@ func TestConcurrentRefresh(t *testing.T) {
...
+	b := &fakeCerts{}
+	certSource := blockingCertSource{
+		map[string]*fakeCerts{}}
+	c := &Client{
+		Certs: &certSource,
+		Dialer: func(string, string) (net.Conn, error) {
+			atomic.AddUint64(&dials, 1)
+
+			// Wait a second to ensure the max connections count is reached by concurrent dialers
+			time.Sleep(time.Second)
```
Comment (Carrotman42) [inline proxy/proxy/client_test.go]: Sleeping in a test is just asking for a flaky test in the long run unfortunately. I think the Dial only needs to block until after the first failure to dial, i.e. dials can unblock after the first `c.handleConn` returns. Feel free to pick how you implement, but I suggest something like:

    firstDialExited := make(chan struct{})
    c := &Client{
       ...,
       Dialer: func(string, string) (net.Conn, error) {
          atomic.AddUint64(&dials, 1)
          <-firstDialExited
          return nil, errFakeDial
       },
    }
    ...
    var wg sync.WaitGroup
    var firstDialOnce sync.Once
    for _, instanceName := range instanceNames {
       wg.Add(1)
       go func(instanceName string) {
          defer wg.Done()

          conn := Conn{...}
          c.handleConn(conn)

          firstDialOnce.Do(func() { close(firstDialExited) }
       }(instanceName)
    }

wdyt?

File: proxy/proxy/client.go (hunk 6 above)
Comment (pdecat) [inline proxy/proxy/client.go]: Indeed, fixed.

File: proxy/proxy/client_test.go (hunk 7 above)
Comment (pdecat) [inline proxy/proxy/client_test.go]: Done.

File: proxy/proxy/client_test.go (hunk 8 above)
Comment (pdecat) [inline proxy/proxy/client_test.go]: Yup, the sleep is ugly and I was looking for other ways avoiding it too. trying to use a WaitGroup in the Dialer but this was going nowhere...

Your proposal works great and is way simpler, go for it :)

**Comment by Carrotman42** [comment]
Cool, unit test lgtm.

Are you able to add an end-to-end test in this pull request? If you're willing to do some refactoring in `tests/gce_test.go` which put the setup logic from TestGCE into a subroutine and was reused between the existing TestGCE and a new TestConnectionLimit that'd be nice. Alternatively you could write a new test in `tests/dialers_test.go` which calls `proxy.InitClient` to re-initialize the default client and use the `mysql.Dial` hook to verify that only a certain number connections are available.

I'd prefer the `gce_test.go` version but that is a little more work for you, so feel free to have the end-to-end test via the simpler "dialers" package.

**Comment by pdecat** [comment]
I glanced at those two tests yesterday and also thought `gce_test.go` was the way to go.
+1 to refactor reusable parts.

One thing that bugged me is how do you automate those tests?
Wouldn't it better to also create the CloudSQL instance during the test?

**Comment by Carrotman42** [comment]
We don't actually automatically run these tests, there are other tests that
we regularly run inside Google instead. I mostly added these as an
after-thought in case people outside Google wanted to add them to their CI
for any reason, but Im not entirely sure why I didn't right the code to
automatically create the SQL database as well. It's probably that you can't
reuse database instance names for a few days after deleting it (there's an
internal bug about it but it's been an open issue for a while). I run these
tests when I do a Proxy release and I just manually create a small instance
first before running the tests.

If you want to add a helper library for creating a Database instance feel
free, but it'd definitely be easier for you to just make a small database
for testing manually.

**Comment by pdecat** [comment]
Hi @Carrotman42, I had to resort to using sleeps to implement the e2e test as we have way less control on what can be done on the remote mysql client.

Probably less than an issue in those tests than in the unit tests, what do you think?

File: tests/connection_limit_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+	cmd := fmt.Sprintf(`mysql -uroot -S cloudsql/%s -e "SELECT 1; SELECT SLEEP(120);"`, *databaseName)
+
+	// Use less than the sshd MaxStartups configuration (defaults to 10)
+	for i := 0; i < 5; i++ {
+		go func() {
+			log.Print("Starting blocking mysql command")
+			var sout, serr bytes.Buffer
+			if err = sshRun(ssh, cmd, nil, &sout, &serr); err != nil {
+				t.Fatalf("Error running mysql: %v\n\nstandard out:\n%s\nstandard err:\n%s", err, &sout, &serr)
```
[the remark this answers is not in the document]
Comment (pdecat) [reply tests/connection_limit_test.go]: Good catch, fixed.

**Comment by pdecat** [comment]
@Carrotman42 rebased and conflicts resolved.

File: tests/connection_limit_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (AthenaShi) [reply tests/connection_limit_test.go]: @pdecat I might miss something, but I don't see this changed.

[the remark this answers is not in the document]
Comment (pdecat) [reply tests/connection_limit_test.go]: @AthenaShi oops, you're right, I forgot to push this change done on another workspace. Fixed!
